kernel/aster-nix/src/fs/utils/page_cache.rs

```diff
@@ -85,16 +87,193 @@ impl Debug for PageCache {
     }
 }
 
+struct ReadaheadWindow {
+    /// The window.
+    window: Range<usize>,
+    /// Look ahead position in the current window, where the readahead is triggered.
+    /// TODO: We set the lookahead_index to the start of the window for now.
+    /// This should be adjustable by the user.
+    lookahead_index: usize,
+}
+
+impl ReadaheadWindow {
+    pub fn new(window: Range<usize>) -> Self {
+        let lookahead_index = window.start;
+        Self {
+            window,
+            lookahead_index,
+        }
+    }
+
+    pub fn next(&self, max_size: usize, max_page: usize) -> Self {
+        let new_start = self.window.end;
+        let cur_size = self.window.end - self.window.start;
+        let new_size = (cur_size * 2).min(max_size).min(max_page - new_start);
+        Self {
+            window: new_start..(new_start + new_size),
+            lookahead_index: new_start,
+        }
+    }
+
+    pub fn lookahead_index(&self) -> usize {
+        self.lookahead_index
+    }
+
+    pub fn readahead_index(&self) -> usize {
+        self.window.end
+    }
+
+    pub fn ra_range(&self) -> Range<usize> {
+        self.window.clone()
+    }
+}
+
+struct ReadaheadState {
+    /// Current readahead window.
+    ra_window: Option<ReadaheadWindow>,
+    /// Maximum window size.
+    max_size: usize,
+    /// Last page visited.
+    prev_page: Option<usize>,
+    /// Readahead requests waiter.
+    waiter: BioWaiter,
+}
+
+impl ReadaheadState {
+    const INIT_WINDOW_SIZE: usize = 4;
+    const DEFAULT_MAX_SIZE: usize = 32;
+
+    pub fn new() -> Self {
+        Self {
+            ra_window: None,
+            max_size: Self::DEFAULT_MAX_SIZE,
+            prev_page: None,
+            waiter: BioWaiter::new(),
+        }
+    }
+
+    /// Set the maximum readahead window size.
+    pub fn set_max_window_size(&mut self, size: usize) {
+        self.max_size = size;
+    }
+
+    fn is_sequential(&self, idx: usize) -> bool {
+        if let Some(prev) = self.prev_page {
+            idx == prev || idx == prev + 1
+        } else {
+            false
+        }
+    }
+
+    /// The number of uncompleted requests in waiter.
+    /// This number will be zero if there are no previous readahead.
+    pub fn request_number(&self) -> usize {
+        self.waiter.nreqs()
+    }
+
+    fn has_completed_all_requests(&self) -> bool {
```

Comment:
The code in Line 191`(self.has_completed_all_requests() && self.request_number() != 0)` can be move into this fuction:
```rust
 fn has_completed_all_requests(&self) -> bool {
     let nreqs = self.request_number();
     if nreqs == 0 {
        return false;
     }

     for i in 0..self.nreqs {
            if self.waiter.status(i) == BioStatus::Submit {
                return false;
            }
     }
     true
}
```